Sources/Conduit/Networking/Data/DataDownloader.swift

```diff
@@ -0,0 +1,141 @@
+    public init(cache: URLDataCache,
+                sessionClient: URLSessionClientType = URLSessionClient(),
+                completionQueue: OperationQueue? = nil) {
+        self.cache = cache
+        self.sessionClient = sessionClient
+        self.completionQueue = completionQueue
+    }
+
+    /// Downloads data or retrieves it from the cache if previously downloaded.
+    /// - Parameters:
+    ///     - request: The request for the data
+    /// - Returns: A concrete SessionTaskProxyType
+    @discardableResult
+    public func downloadData(for request: URLRequest, completion: @escaping CompletionHandler) -> SessionTaskProxyType? {
+        var proxy: SessionTaskProxyType?
+        let completionQueue = self.completionQueue ?? .current ?? .main
+
+        serialQueue.sync { [weak self] in
+            guard let `self` = self else {
+                return
+            }
+
+            if let data = self.cache.data(for: request) {
+                let response = Response(data: data, error: nil, urlResponse: nil, isFromCache: true)
+                completion(response)
+                return
+            }
+
+            guard let cacheIdentifier = self.cache.cacheIdentifier(for: request) else {
+                let response = Response(data: nil,
+                                        error: DataDownloaderError.invalidRequest,
+                                        urlResponse: nil,
+                                        isFromCache: false)
+                completion(response)
+                return
+            }
+
+            self.register(completionHandler: completion, for: cacheIdentifier)
+
+            if let sessionTaskProxy = self.sessionProxyMap[cacheIdentifier] {
+                proxy = sessionTaskProxy
+                return
+            }
+
+            // Strongly capture self within the completion handler to ensure
+            // DataDownloader is persisted long enough to respond
+            let strongSelf = self
+            proxy = self.sessionClient.begin(request: request) { data, response, error in
+                if let data = data {
+                    strongSelf.cache.cache(data: data as NSData, for: request)
+                }
+
+                let response = Response(data: data as NSData?, error: error, urlResponse: response, isFromCache: false)
+
+                func execute(handler: @escaping CompletionHandler) {
+                    completionQueue.addOperation {
+                        handler(response)
+                    }
+                }
+
+                // Intentional retain cycle that releases immediately after execution
+                strongSelf.serialQueue.async {
+                    strongSelf.sessionProxyMap[cacheIdentifier] = nil
+                    strongSelf.completionHandlerMap[cacheIdentifier]?.forEach(execute)
+                    strongSelf.completionHandlerMap[cacheIdentifier] = nil
+                }
```

Comment:
Could you please explain the need for intentional retain cycles a bit more?

Reply:
@bconway99 If you look at the comment above that
```
// Strongly capture self within the completion handler to ensure 
// DataDownloader is persisted long enough to respond
```
Since there are multiple asynchronous elements in this method we want to make sure we set the data we need before the objects are released